### PR TITLE
fix: derank peers that responded with bad data

### PR DESCRIPTION
**Pull request by mattsse**
alternative to #7849

this mitigates an issue in live sync, or more general, when the issuer of the request does not penalize an empty response, in which case we would continue sending followups to the same peer if it has the lowest rtt.

this keeps track of the last returned response and uses it when selecting peers.
peers that responded with empty data or an error are de-prioritized

### CODE REVIEW

File: crates/net/network/src/fetch/mod.rs [hunk 1]
```diff
@@ -119,14 +127,27 @@ impl StateFetcher {
...
+            if best_peer.1.last_response_likely_bad && !maybe_better.1.last_response_likely_bad {
+                best_peer = maybe_better;
+                continue
+            }
+
+            if maybe_better.1.timeout() < best_peer.1.timeout() &&
+                !maybe_better.1.last_response_likely_bad
+            {
+                best_peer = maybe_better;
+            }
```
Comment (onbjerg) [inline crates/net/network/src/fetch/mod.rs:147]: nit: add small note on both these conds

```suggestion
            // replace best peer if our current best peer sent us a bad response last time
            if best_peer.1.last_response_likely_bad && !maybe_better.1.last_response_likely_bad {
                best_peer = maybe_better;
                continue
            }

            // replace best peer if this peer has better rtt
            if maybe_better.1.timeout() < best_peer.1.timeout() &&
                !maybe_better.1.last_response_likely_bad
            {
                best_peer = maybe_better;
            }
```

**Review by onbjerg — APPROVED** [review]
this is less invasive

**Comment by joshieDo** [comment]
Tried it, works but I feel like there's some nuances to be aware of.

**Over time**, we will keep prioritizing higher latency peers, because we tag the lower latency ones with `last_response_likely_bad`.  In the scenario where all are tagged, the issue will remain where the peer with the lowest latency will be asked over and over, if it so happens that it returns an empty response.

But maybe this doesn't happen, since this path may not be taken many times? Maybe in some E2E convoluted scenario .

File: crates/net/network/src/fetch/mod.rs (hunk 1 above)
Comment (joshieDo) [reply crates/net/network/src/fetch/mod.rs:147]: > if our current best peer sent us a bad response last time

would replace with

> if our current best peer sent us a bad response in the past

**Review by joshieDo — APPROVED** [review]

**Pull request merged**